game-core/src/main/java/games/strategy/engine/data/GameMap.java

```diff
@@ -271,80 +272,110 @@ public Territory getTerritory(final String s) {
   }
 
   /**
-   * Returns the shortest route between two territories or null if no route exists.
+   * Returns all neighbors within a certain distance of the starting territory that match the
+   * condition. Does NOT include the original/starting territory in the returned Set.
    *
-   * @param t1 start territory of the route
-   * @param t2 end territory of the route
+   * <p>TODO: update to properly consider movement cost not just distance
+   *
+   * @param territory referring territory
+   * @param movementLeft maximal distance of the neighboring territories
    */
-  public Route getRoute(final Territory t1, final Territory t2) {
-    return getRoute(t1, t2, Matches.territoryIsLandOrWater());
+  public Set<Territory> getNeighborsByMovementCost(
+      final Territory territory, final Unit unit, final BigDecimal movementLeft) {
+    Preconditions.checkArgument(
+        movementLeft.compareTo(BigDecimal.ZERO) >= 0,
+        "MovementLeft must be non-negative: " + movementLeft);
+    if (movementLeft.compareTo(BigDecimal.ZERO) == 0) {
+      return Collections.emptySet();
+    }
+    final Set<Territory> start = getNeighbors(territory);
+    if (movementLeft.compareTo(BigDecimal.ONE) <= 0) {
```

Comment:
Is this an impossible condition? It was asserted movement left is non-negative at the start of the method, and the previous if statement checks for zero movement left, which should leave no case for zero or less. Is there something I'm overlooking?

The logic seems complex enough that it would be a good candidate for test. That would help ensure no bugs and give us confidence for any future updates that we are not breaking the method and save a lot to understand every case. `GameMap` perhaps is also not the best home for every calculation relating to game map, to test this one would need to control the `getNeighbors` function, and it would be straight forward. To do that, a utility class with a passed in function to calculate neighbors could be mocked or stubbed, making testing pretty straight forward.

Reply:
This is comparing to 1. So movementLeft could be say 0.5.

Reply:
Ah, I see. It's just a bit difficult to grok: `movementLeft.compareTo(BigDecimal.ONE) <= 0`
Takes more effort to understand than would be really desirable. If not resorting to using doubles, then extracting the condition to a boolean variable so it is named could reduce the overhead to understand the condition.

Reply:
Given how often comparisons are used, its probably overkill as we are probably talking about 100 condition extraction for a very simple comparison.

Reply:
Does this PR really contain around 100 condition checks? From what I've seen it's about a  dozen. 

To be direct, this is hard to understand:
```
if (movementLeft.compareTo(BigDecimal.ONE) <= 0) {
```
It's not impossible, nor particularly difficult, but it's not KISS